src/bin/pg_autoctl/monitor.c

```diff
@@ -1793,6 +1922,171 @@ monitor_formation_uri(Monitor *monitor, const char *formation,
 }
 
 
+/*
+ * monitor_print_every_formation_uri prints a table of all our connection
+ * strings: first the monitor URI itself, and then one line per formation.
+ */
+bool
+monitor_print_every_formation_uri(Monitor *monitor)
+{
+	FormationURIParseContext context = { 0 };
+	PGSQL *pgsql = &monitor->pgsql;
+	const char *sql =
+		"SELECT 'monitor', 'monitor', $1 "
+		" UNION ALL "
+		"SELECT 'formation', formationid, formation_uri "
+		"  FROM pgautofailover.formation, "
+		"       pgautofailover.formation_uri(formation.formationid)";
+
+	int paramCount = 1;
+	Oid paramTypes[1] = { TEXTOID };
+	const char *paramValues[1];
+
+	paramValues[0] = monitor->pgsql.connectionString;
+
+	context.parsedOK = false;
+
+	if (!pgsql_execute_with_params(pgsql, sql,
+								   paramCount, paramTypes, paramValues,
+								   &context, &printFormationURI))
+	{
+		log_error("Failed to list the formation uri, "
+				  "see previous lines for details.");
+		return false;
+	}
+
+	if (!context.parsedOK)
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* disconnect from PostgreSQL now */
+	pgsql_finish(&monitor->pgsql);
+
+	return true;
+}
+
+
+/*
+ * monitor_print_every_formation_uri_as_json prints all our connection strings
+ * in the JSON format: first the monitor URI itself, and then one line per
+ * formation.
+ */
+bool
+monitor_print_every_formation_uri_as_json(Monitor *monitor, FILE *stream)
+{
+	SingleValueResultContext context = { { 0 }, PGSQL_RESULT_STRING, false };
+	PGSQL *pgsql = &monitor->pgsql;
+	const char *sql =
+		"WITH formation(type, name, uri) AS ( "
+		"SELECT 'monitor', 'monitor', $1 "
+		" UNION ALL "
+		"SELECT 'formation', formationid, formation_uri "
+		"  FROM pgautofailover.formation, "
+		"       pgautofailover.formation_uri(formation.formationid)"
+		") "
+		"SELECT jsonb_pretty(jsonb_agg(row_to_json(formation))) FROM formation";
+
+	int paramCount = 1;
+	Oid paramTypes[1] = { TEXTOID };
+	const char *paramValues[1];
+
+	paramValues[0] = monitor->pgsql.connectionString;
+
+	if (!pgsql_execute_with_params(pgsql, sql,
+								   paramCount, paramTypes, paramValues,
+								   &context, &parseSingleValueResult))
+	{
+		log_error("Failed to list the formation uri, "
+				  "see previous lines for details.");
+		return false;
+	}
+
+	if (!context.parsedOk)
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* disconnect from PostgreSQL now */
+	pgsql_finish(&monitor->pgsql);
+
+	fprintf(stream, "%s\n", context.strVal);
```

Comment:
context.strVal is not deallocated

Reply:
I believe we are going to go over all these in a separate work. so you can ignore this one